New blog post - RIMdev Radio: Building with Astro (#483)
ritterim/ritterim.github.io: _posts/2022-10-12-rimdev-radio-building-with-astro.md

date: 2022-10-12 12:00:00
tags:
- Astro
- SCSS
- Supabase
- Frontend
- Spotify
- Netlify
categories:
- CSS
- JavaScript
- Static Site Generators
twitter_text: "Builing a dynamic site with Astro."
authors: Ryan Trimble
image: https://images.unsplash.com/photo-1505740420928-5e560c06d30e?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80
image_url: https://unsplash.com/photos/1oKxSKSOowE
image_credit: Mohammad Metri
---

My current favorite thing in web dev is a new site generator called [Astro](https://astro.build/). 

Astro is like [Hugo](https://gohugo.io/) and [11ty](https://11ty.dev), yet contains a few neat tricks right out of the box. The main goal of Astro is to deliver as little JavaScript to the browser as possible. It handles this by running JavaScript in the build process or when rendered server-side. That is not to say you can't run JavaScript in the browser though. In fact, you can include components from any of the major JS frameworks, such as Vue, React, and Svelte. This is possible through a concept Astro is pioneering: [component islands](https://docs.astro.build/en/concepts/islands/).

> I recently [delivered a talk](https://astro-friday-talk-slides.netlify.app/1) on Astro to our dev team as part of the "Friday Talks" we do here in RIMdev. I demonstrated how to build with Astro and it's server-side rendering capability. The [first demo](https://astro-friday-talk.netlify.app/) shows off it's framework agnostic abilities, and how to run JavaScript. The [second demo](http://astro-pokemon.netlify.app/) shows how you can run JavaScript server-side and deliver only HTML. These projects were fun to build, but I wanted to try and build something with a bit more substance.

Reviewer: ```suggestion
I recently [delivered a talk](https://astro-friday-talk-slides.netlify.app/1) on Astro to our dev team as part of the "Friday Talks" we do here in RIMdev. I demonstrated how to build with Astro and it's server-side rendering capability. The [first demo](https://astro-friday-talk.netlify.app/) shows off its framework agnostic abilities and how to run JavaScript. The [second demo](http://astro-pokemon.netlify.app/) shows how you can run JavaScript server-side and deliver only HTML. These projects were fun to build, but I wanted to try and build something with a bit more substance.
```

The idea was that team members would be able to search Spotify for songs and update the playlist. Recommendations get stored in a database table, which would then display on the home page. The idea didn't quite pan out due to limitations of the Spotify API, but still able to deliver with a slight pivot.

In Spotify, you can set up collaborative playlists that allows you and friends to add songs to a playlist. This function is not available to the API though, as spammers could overrun playlists. I learned this at the finish line, so while I can update the playlist through the API - no one else can. Unfortunate as this was the main feature of the app. 

That's OK though! Team members that choose to take part can send me songs they would like included and I can update the playlist. This was a super fun weekend project and I was able to learn a lot, so no harm done!

## Piecing it all together

### Astro

Everything happens inside Astro and luckily getting started is quite easy!

The `npm create astro@latest` command walks you through the process of setting up a new instance. I choose the default options for everything, except opting out of using TypeScript. 

> From there, the new site will scaffold and there are few organizational things to take note of:

Reviewer: ```suggestion
From there, the new site will scaffold and there are a few organizational things to take note of:
```

- The **Pages** directory is where each site page will live. These will use the `.astro` file extension, and are capable of running JavaScript at build.  So you can generate HTML code by running certain JavaScript functionality.
- The **Components** directory will house UI components. These can be Astro components or components for whichever framework you prefer.
- The **Layouts** directory is where you can include components used for layout purposes.

As I intended to use Sass for styling, I also added a `styles` folder to the `src` directory.

> Speaking of which, to setup Sass - you do need to install Sass as a dependency to use it:

Reviewer: ```suggestion
Speaking of which, to set up Sass - you do need to install Sass as a dependency to use it:
```

    .then((data) => {
      options.value = [...data.tracks.items];
    })
    .catch((err) => console.log(err));
};
```

I set up a separate API call which adds items to a playlist, but as previously mentioned, does *not* allow users to update playlists they do not own.

### Netlify

Netlify is a great service for deploying static sites. Like Amazon Web Service's product Lambda, Netlify also provides the capability of running cloud functions. Astro can be built in a way to leverage this feature, enabling server-side rendering. This may sound complicated at first, but it is almost *too* easy as Astro provides an integration to help set this up.

```
npx astro add netlify
```

Running the add Netlify integration command will automatically install necessary packages and update the `astro.config.mjs` file.

Now all that is needed is to deploy the site to Netlify. I typically do this by connecting Netlify to the GitHub repo, which gives the added benefit of automatically re-deploying the site when pushing to the main branch. You can also use PR Previews to see changes to the site before a pull request is even merged to the main branch.

## Final Thoughts

> Overall working with Astro is a super enjoyable experience. I'm looking forward to building more projects with it!

Reviewer: ```suggestion
Overall, working with Astro is a super enjoyable experience. I'm looking forward to building more projects with it!
```
